src/mapgen.cpp

```diff
@@ -193,44 +194,153 @@ void mapgen_function_builtin::generate( mapgendata &mgd )
 ///// mapgen_function class.
 ///// all sorts of ways to apply our hellish reality to a grid-o-squares
 
+class mapgen_basic_container
+{
+    private:
+        std::vector<std::shared_ptr<mapgen_function>> mapgens_;
+        weighted_int_list<std::shared_ptr<mapgen_function>> weights_;
+
+    public:
+        int add( const std::shared_ptr<mapgen_function> ptr ) {
+            assert( ptr );
+            mapgens_.push_back( ptr );
+            return mapgens_.size() - 1;
+        }
+        void erase( const size_t index ) {
+            if( index > mapgens_.size() ) {
+                return;
+            }
+            assert( mapgens_[index] );
+            // Can't actually erase it as this might invalid the index stored elsewhere
+            mapgens_[index]->weight = 0;
+        }
+        /**
+         * Pick a mapgen function randomly.
+         * @p hardcoded_weight Weight for an additional entry. If that entry is chosen,
+         * a null pointer is returned. If unsure, just use 0 for it.
+         */
+        mapgen_function *pick( const int hardcoded_weight ) const {
```

Comment:
Shouldn't this function return a constant pointer? The function itself is marked `const`, but its result points at the class internals which can be changed through it.

```suggestion
        const mapgen_function *pick( const int hardcoded_weight ) const {
```

Reply:
Yes, it should. But `mapgen_function::generate` is not const either, so you would have to cast the pointer anyway in order to call that function. And calling `generate` is basically the only thing one will do with that pointer.

Reply:
`generate` should probably me marked const either. It's meant to operate on `mapgendata` and shouldn't mutate its internal state. Anyways, a const function which returns a non-const pointer to its internals is suspicious at the very least.